pkg/virtual/apiexport/controllers/apireconciler/apiexport_apireconciler_reconcile.go

```diff
@@ -57,29 +63,66 @@ func (c *APIReconciler) reconcile(ctx context.Context, apiExport *apisv1alpha1.A
+	// schemaIdentity will map the schema to the given API export that it comes from.
+	// and potentially a
+	apiResourceSchemas, schemaIdentity, err := c.getSchemasFromAPIExport(apiExport)
+	if err != nil {
+		return err
 	}
 
-	for _, schemaName := range apiExport.Spec.LatestResourceSchemas {
-		apiResourceSchema, err := c.apiResourceSchemaLister.Get(clusters.ToClusterAwareKey(logicalcluster.From(apiExport), schemaName))
-		if err != nil && !apierrors.IsNotFound(err) {
-			return err
+	for _, pc := range apiExport.Spec.PermissionClaims {
+		internal, schema := isPermissionClaimForInternalAPI(pc)
+		if internal {
+			shallow := *schema
+			shallow.ClusterName = logicalcluster.From(apiExport).String()
+			apiResourceSchemas = append(apiResourceSchemas, &shallow)
+			schemaIdentity[shallow.Name] = exportAndClaimInfo{}
+			continue
 		}
-		if apierrors.IsNotFound(err) {
-			klog.V(3).Infof("APIResourceSchema %s in APIExport %s|% not found", schemaName, apiExport.Namespace, apiExport.Name)
+		if !internal && pc.IdentityHash != "" {
+			exports, err := c.apiExportIndexer.ByIndex(indexers.IndexAPIExportByIdentity, pc.IdentityHash)
+			if err != nil {
+				return err
+			}
+			if len(exports) != 1 {
+				return fmt.Errorf("got %d api exports expected one", len(exports))
+			}
+			export := exports[0]
+			// For the export that si found based on an the identity hash
+			// we need to add all the latest schemas that the export provides
+			// We all need to map the schema to the given APIExport that it comes from
+			// so that when creating the api definition for it, we can use the given
+			// api export to fully fill in and serve the API for the virtual workspace.
+			owningExport, ok := export.(*apisv1alpha1.APIExport)
+			if !ok {
+				return fmt.Errorf("invalid type from apiExportIndexer got: %T, expected %T", export, apiExport)
+			}
+			schemas, schemaIDs, err := c.getSchemasFromAPIExport(owningExport)
+			if err != nil {
+				return err
+			}
+			for _, schema := range schemas {
+				if schema.Spec.Group != pc.Group || schema.Spec.Names.Plural != pc.Resource {
+					continue
+				}
+				apiResourceSchemas = append(apiResourceSchemas, schema)
+			}
+			for k := range schemaIDs {
+				// Override the permission claims that should be used. for this schema identity
+				// Here we will set the identity for the schema for a permission claim.
+				exportAndClaim := schemaIdentity[k]
+				exportAndClaim.claim = &pc
+				exportAndClaim.identityHash = owningExport.Status.IdentityHash
+				schemaIdentity[k] = exportAndClaim
+			}
 			continue
 		}
-		apiResourceSchemas = append(apiResourceSchemas, apiResourceSchema)
-		schemaIdentity[schemaName] = apiExport.Status.IdentityHash
+		// TODO: Add a openapi validation to catch this case, and it should never happen
+		klog.Warningf("Permission claim %v for APIEXport %s is not internal and does not have an identity hash",
+			pc,
+			fmt.Sprintf("%s|%s", logicalcluster.From(apiExport), apiExport.Name),
```

Comment:
(style) unusual factoring. Everywhere else we do `klog.Warningf(".... %s|%s ....", logicalcluster.From(apiExport), apiExport.Name)`. This extra sprintf is strange.

Reply:
done